First cut at a foo_image rule

mattmoor opened the pull request:
This change is a first-cut at attempting a `foo_image` rule that directly assembles a Docker image compatible with `foo_binary`, but leveraging a sharded archive format (vs. a mono-archive format like [PAR](https://github.com/google/subpar) or "deploy JAR")

The advantage of the sharded archive format is simple:  **We want to translate build-time incrementality into redeployment incrementality.**

By enabling the partitioning of dependencies into their own layers, in a binary agnostic fashion, we enable:
1. sharing of those dependencies across different images that depend on them
1. no republishing of those dependencies' layers when they haven't changed
1. no repulling unchanged dependencies' layers

In this model, what gets repushed/repulled can be made a direct function of Bazel's library-level incrementality.

mattmoor commented:
@duggelz I can't make you a reviewer (thanks, GitHub), but I'd very much appreciate your feedback on this.  I'm sure I'm missing tons of stuff (e.g. the `empty_filenames` fun) that you've surely dealt with in `subpar`.

mattmoor commented:
@damienmg I have no idea why the Bazel presubmit is unhappy, when I click through the `!`s everything is green...

damienmg commented:
build_test is failing on all the config:

http://ci.bazel.io/blue/rest/organizations/jenkins/pipelines/PR/rules_docker/runs/66/nodes/14/log/?start=0

On Fri, Jul 21, 2017, 8:02 PM mattmoor <notifications@github.com> wrote:

> @damienmg <https://github.com/damienmg> I have no idea why the Bazel
> presubmit is unhappy, when I click through the !s everything is green...
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/rules_docker/pull/98#issuecomment-317071513>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADjHfyJeIhc7_fDlsAQK8Y8PbqxAnfE_ks5sQOe-gaJpZM4OftjJ>
> .
>

mattmoor commented:
Thanks @damienmg.  Gotta say, I kinda hate the new Jenkins UI :-/  Don't show me green for the step that failed, so I need to expand every stupid thing. </grumble>

mattmoor commented:
I added `cc_image` too.

I debated adding `go_image` too, but I think adding tests would and up requiring the `go_prefix` stuff, which might make `rules_go` a harder dependency than I'd like.

I'll experiment with JAR / WAR rules once I'm back from vacation.

mattmoor commented:
@damienmg thoughts?

duggelz commented on docker/testdata/BUILD:
```diff
@@ -344,13 +344,13 @@ load(
-   stamp = True
+    name = "bundle_to_push",
+    images = {
+        "us.gcr.io/convoy-adapter/bazel-test/{BUILD_USER}:foo": ":with_env",
```
It looks like this is both a functionality test, and implicitly a system test of the gcr regions (or whatever they're called).  Can you explicitly document that this is "a docker image from each GCR region" so that it's clear that we need to add a "moon.gcr.io" when we open the lunar base?

duggelz commented on docker/testdata/py_image.py:
```diff
@@ -0,0 +1,25 @@
+from docker.testdata import py_image_library
+
+def main():
+  print 'First: %d' % py_image_library.fn(1)
```
Can you make these print statements Python3-compatible?

duggelz commented on testing/e2e.sh:
```diff
@@ -62,6 +62,77 @@ EOF
+git_repository(
+    name = "io_bazel_rules_go",
+    remote = "https://github.com/bazelbuild/rules_go.git",
+    tag = "0.4.4",
```
We should either collect all the pinned versions/hashes in a document somewhere, or have a janitor-bot remind us to update them on a regular basis.

damienmg replied:
I know :( unfortunately it is not on our roadmap yet. I think we should build some kind of tools that would update those versions callable from bazel e.g. `bazel tools bump-dep-versions` o

mattmoor commented on docker/contrib/python/image.bzl:
```diff
@@ -0,0 +1,171 @@
+_dep_layer = rule(
+    attrs = _build_attrs + {
+        # The base image on which to overlay the dependency layers.
+        "base": attr.label(default = Label("@py_image_base//image")),
```
That's a good question.

For the latter, I think this is just up to the toolchain that `rules_go` pulls in, and I punt on importing the Go rules in this repo (they must be imported before `contrib/go/image.bzl` or it will fail).

For the former, I haven't done anything specifically to address this, but I think the story should be one of:
1. Change the `base` to an appropriate image (requires exposing `base`, which has a TODO now)
1. Make the rules smart about `PY2`, `PY3`, `PY2AND3` in its choice of the distroless base (and perhaps override-able by the above) .

I would love to hear your thoughts and/or clever ideas here.

mattmoor commented on docker/contrib/cc/image.bzl:
```diff
@@ -0,0 +1,173 @@
+    directory="/app/" + ctx.label.package,
+    files=list(ctx.attr.dep.default_runfiles.files),
+    symlinks={
+      # Handle empty files by linking to /dev/null
```
I added a comment for now linking to your issues, this is a great catch thanks.

mattmoor commented on docker/contrib/cc/image.bzl:
```diff
@@ -0,0 +1,173 @@
+    for input in available
+  } + {
+    # Handle empty files by linking to /dev/null
+    base_directory + "/" + empty: "/dev/null"
```
Ack

mattmoor commented:
I pushed support for `war_image`, which is mildly different.  I am going to return to experimenting with `java_image`, for which I may end up refactoring some of this new code to share.

This is largely based on [this](https://github.com/bazelbuild/rules_appengine/blob/master/appengine/appengine.bzl#L130-L139), since I'm no WAR expert...

duggelz commented on docker/contrib/python/image.bzl:
```diff
@@ -0,0 +1,171 @@
+    # image is `docker run ...`.
+    # Per: https://docs.docker.com/engine/reference/builder/#entrypoint
+    # we should use the "exec" (list) form of entrypoint.
+    entrypoint=['/usr/bin/python', binary_name],
```
You're right on both counts, I wasn't thinking.

damienmg commented on docker/build.bzl:
```diff
@@ -68,23 +102,26 @@ def _build_layer(ctx):
-           for k in ctx.attr.symlinks]
+  for k in symlinks:
+    if ':' in k:
+      fail("The source of a symlink cannot contain ':', got: %s" % k)
```
why?

mattmoor replied:
Because of how we pass it to the underlying binary.
  "foo:bar": "baz"

Would be passed indistinguishably from:
  "foo": "bar:baz"

damienmg commented on docker/contrib/cc/image.bzl:
```diff
@@ -0,0 +1,60 @@
+      digest = "sha256:942eb947818e7e32200950b600cc94d5477b03e0b99bf732b4c1e2bba6eec717",
+    )
+
+def cc_image(name, base=None, deps=[], layers=[], **kwargs):
```
cc_docker_build to be consistent in naming?

mattmoor replied:
I think I prefer the current naming given it's current home under `rules_docker`.  `docker_image` probably would have been a better name :)

damienmg commented on docker/contrib/go/image.bzl:
```diff
@@ -0,0 +1,64 @@
+      digest = "sha256:06fcd3edcfeefe13b82fa8bdb9e3f4fa3bf4c7e8fe997bee0230e392f77d0e04",
+    )
+
+def go_image(name, base=None, deps=[], layers=[], **kwargs):
```
idem go_docker_build

damienmg commented on docker/build.bzl:
```diff
@@ -11,7 +11,35 @@
+
+  load(
+    "@io_bazel_rules_docker//docker:build.bzl",
+    _build_attrs="attrs",
```
That would be good to respect the naming we do native sandwich.

E.g. that would be:
_docker="docker"

docker would then be a struct inside build.bzl:
docker = struct(
  build = struct(
      impl=_implementation,
      attrs=_attrs,
      outputs=_outputs
  )
)

And you can call _docker.build.impl directly. Also we might want to switch to [named provider](https://bazel.build/designs/skylark/declared-providers.html) if we start offering interaction API

mattmoor replied:
It took me a little while to understand what you were suggesting with the structs, but I still have no idea what you are suggesting with the providers.  It looks to me like Bazel has reinvented proto options, but I'm not sure how that helps here.

damienmg replied:
There is an internal doc (/cc @laurentlb for making that doc public) about cleaning up namespace of native symbols. A consequence would be that we would write java.merge instead of java_common.merge, with `java` being a symbol loaded from an extension. That's the reason for struct.

For named providers this is just something to consider since this should be the new way.

mattmoor commented:
@damienmg Any suggestions for how I add empty files at arbitrary paths?

duggelz commented:
Re: symlinks to empty files, the Java code inside Bazel handles this with a Map where entries have the "symlink target" key as whatever, and the "symlink source" value is NULL.  So maybe try constructing the symlink dictionary with entries like `'myfilename': None`, and see if that None survives all the way from Skylark to Bazel to the sandboxer.

https://github.com/bazelbuild/bazel/commit/738e5727bab23961c7cc423606f38f8b97a8150e#diff-d336e8f8c32246f283d163ad11f0c40dR106

mattmoor commented:
@duggelz the symlink dictionary we use directs a custom script (`build_tar.py`, currently forked from `pkg_tar`) on how to create symlinks within the resulting tarball (it isn't for Bazel's symlink forest).

That said, we can probably plumb through a list of empty filenames for it to create.  I'll look at that next.


FWIW, I just fixed a bug in `java_image` and modified the test to use an external dependency to cover it.

mattmoor commented:
Ok, we no longer symlink `/dev/null` for empty files.

mattmoor commented:
I'm going to merge this to checkpoint, as I already have at least one follow on change, and I expect to find bugs as folks start to use this.  Feel free to continue to comment if you find any issues and I'll send PRs to fix.

mmikulicic commented on docker/contrib/go/image.bzl:
```diff
@@ -0,0 +1,64 @@
+  binary_name = name + ".binary"
+
+  if layers:
+    print("go_image does not benefit from layers=[], got: %s" % layers)
```
@mattmoor I'm confused about this warning. I'm I not supposed to pass layers when using go_image?
